### PR TITLE
mam in progress

### PR DESCRIPTION
**Pull request by liuch**

### CODE REVIEW

File: src/xmpp/xmpp-im/types.cpp [hunk 1]
```diff
@@ -1611,34 +1612,51 @@ const IBBData& Message::ibbData() const
     return d->ibbData;
 }
 
-void Message::setDisabledCarbons(bool disabled)
+Jid Message::displayJid() const
```
Comment (Ri0n) [inline src/xmpp/xmpp-im/types.cpp]: please add a comment explaining why it was added.

File: src/xmpp/xmpp-im/types.cpp [hunk 2]
```diff
@@ -1611,34 +1612,51 @@ const IBBData& Message::ibbData() const
...
+Message &Message::displayMessage()
 {
-    d->carbonDir = cd;
+    if (d->forwarding.isCarbons() && d->forwarding.message())
+        return *d->forwarding.message();
+    return *this;
 }
 
-Message::CarbonDir Message::carbonDirection() const
+void Message::setCarbonsPrivate(bool privare)
```
Comment (Ri0n) [inline src/xmpp/xmpp-im/types.cpp]: typo

File: src/xmpp/xmpp-im/xmpp_carbons.cpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+#include <QDomElement>
+
+#include "xmpp_carbons.h"
+#include "xmpp_xmlcommon.h"
+#include "xmpp_message.h"
+#include "xmpp_forwarding.h"
+
+using namespace XMPP;
+
+static const QLatin1String xmlns_carbons("urn:xmpp:carbons:2");
```
Comment (Ri0n) [inline src/xmpp/xmpp-im/xmpp_carbons.cpp]: probably QStringLiteral fit better here

File: src/xmpp/xmpp-im/xmpp_tasks.cpp [hunk 4]
```diff
@@ -891,19 +892,145 @@ void JT_Message::onGo()
...
 
 JT_PushMessage::~JT_PushMessage()
 {
-    delete d;
+}
+
+void JT_PushMessage::subscribeXml(Subscriber *sbs, const QString &tagName, const QString &xmlnsStr, int userData)
+{
+    QString key = d->genKey(tagName, xmlnsStr);
+    QList<Private::SubsData> list = d->subsData.value(key);
```
Comment (pasnox) [inline src/xmpp/xmpp-im/xmpp_tasks.cpp]: You are modifying a copy that would detach on append and cause memory reallocations.
Just use a iterator again for the check.
Then append in this iterator value is it's valid, else call d->subsData.insert(key, {{ sbs, userData }});

File: src/xmpp/xmpp-im/xmpp_tasks.cpp [hunk 5]
```diff
@@ -891,19 +892,145 @@ void JT_Message::onGo()
...
+            return;
+    }
+    list.append({ sbs, userData });
+    d->subsData.insert(key, list);
+}
+
+void JT_PushMessage::unsubscribeXml(Subscriber *sbs, const QString &tagName, const QString &xmlnsStr)
+{
+    QString key = d->genKey(tagName, xmlnsStr);
+    QList<Private::SubsData> list = d->subsData.value(key);
```
Comment (pasnox) [inline src/xmpp/xmpp-im/xmpp_tasks.cpp]: This code is plain wrong.
It remove entries from a temporary list, but the original list is not updated.
Use an iterator and loop it.value() entries to remove them.
Then if it's empty, you can easily remove it with d->subsData.erase(iterator)

**Review by pasnox — CHANGES_REQUESTED** [review]

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+Message *Forwarding::message() const
+{
+    return msg_.get();
+}
+
+void Forwarding::setMessage(const Message &msg)
+{
+    msg_.reset(new Message(msg));
```
Comment (Ri0n) [inline src/xmpp/xmpp-im/xmpp_forwarding.cpp]: all this mixing of smart and regular pointers look so weird. I believe if would be much better to introduce NULL Message concept when Private data is not allocated for the message.
Then regular assignment `msg_ = msg` will be enough.
I'll add relevant changes to the Message class.

Comment (Ri0n) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: pushed

File: src/xmpp/xmpp-im/xmpp_carbons.cpp (hunk 3 above)
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_carbons.cpp]: Agree. Done.

File: src/xmpp/xmpp-im/xmpp_carbons.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+// JT_MessageCarbons
+//----------------------------------------------------------------------------
+JT_MessageCarbons::JT_MessageCarbons(Task *parent)
+    : Task(parent)
+{
+}
+
+void JT_MessageCarbons::enable()
+{
+    iq = createIQ(doc(), QLatin1String("set"), QString::null, id());
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_carbons.cpp]: I didn't know that. Done.

File: src/xmpp/xmpp-im/xmpp_carbons.cpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+void JT_MessageCarbons::onGo()
+{
+    if (!iq.isNull())
+        send(iq);
+}
+
+bool JT_MessageCarbons::take(const QDomElement &e)
+{
+    if (iqVerify(e, Jid(), id())) {
+        if (e.attribute(QLatin1String("type")) != QLatin1String("result"))
```
Comment (pasnox) [inline src/xmpp/xmpp-im/xmpp_carbons.cpp]: There is no QLatin1String variant for QDN::attribute() so it's waonas you domore allocations.
Use QString::fromLatin1("...")

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    }
+    return correct;
+}
+
+QDomElement Forwarding::toXml(Stream *stream) const
+{
+    if (type_ == ForwardedNone || !msg_)
+        return QDomElement();
+
+    QDomElement e = stream->doc().createElement(QLatin1String("forwarded"));
```
Comment (pasnox) [inline src/xmpp/xmpp-im/xmpp_forwarding.cpp]: There is no QDomNode::createElement() taking a QLatin1String, so it's useless as you make 2 allocations, one for QLatin1String and one for QString, just use QString::fromLatin1("...") then.

**Review by pasnox — CHANGES_REQUESTED** [review]
Nearly all QLatin1String usages are plain wrong here.
QLatin1String make sense only if the api you will use has an overload taking a QLatin1String in plus to a QString overload.
If there is only QString parameters and you know you only deal with latin 1 content then just use QString::fromLatin1() then.

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    while (!child.isNull()) {
+        if (child.tagName() == QLatin1String("message")) {
+            if (client->pushMessage()->processXmlSubscribers(child, client, true))
+                break;
+            Stanza s = client->stream().createStanza(addCorrectNS(child));
+            std::unique_ptr<Message> msg(new Message());
+            if (msg->fromStanza(s, client->manualTimeZoneOffset(), client->timeZoneOffset())) {
+                if (client->pushMessage()->processMessageSubscribers(*msg.get(), true))
+                    break;
+                msg_ = std::move(msg);
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: no longer relevant

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+    if (enabled) {
+        d->sbs.reset(new ForwardingSubscriber());
+        d->push_m->subscribeXml(d->sbs.get(), QLatin1String("forwarded"), xmlns_forward, 0);
+        d->push_m->subscribeMessage(d->sbs.get(), 0);
+    }
+    else {
+        d->push_m->unsubscribeXml(d->sbs.get(), QLatin1String("forwarded"), xmlns_forward);
+        d->push_m->unsubscribeMessage(d->sbs.get());
+        d->sbs.release();
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: Probably it was autocompletion. Thanks.

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+{
+    return type_;
+}
+void Forwarding::setType(Type type)
+{
+    if (type_ != type) {
+        type_ = type;
+        if (type == ForwardedNone) {
+            ts_ = QDateTime();
+            ts_.isNull();
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: fixed

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    if (other.msg_.get())
+        msg_.reset(new Message(*other.msg_.get()));
+}
+
+Forwarding::~Forwarding()
+{
+}
+
+Forwarding & Forwarding::operator=(const Forwarding &from)
+{
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: Oh, yes. I forgot that.

File: src/xmpp/xmpp-im/xmpp_forwarding.cpp [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+QDateTime Forwarding::timeStamp() const
+{
+    if (!ts_.isNull())
+        return ts_;
+    return msg_.get() ? msg_->timeStamp() : QDateTime();
+}
+
+void Forwarding::setTimeStamp(QDateTime ts)
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_forwarding.cpp]: Ofcourse.

File: src/xmpp/xmpp-im/xmpp_carbons.cpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    else if (d->sbs.get()) {
+        JT_MessageCarbons *jt = new JT_MessageCarbons(d->push_m->client()->rootTask());
+        connect(jt, &JT_MessageCarbons::finished, this, [=]() {
+            d->enable = false;
+            d->unsubscribe();
+            emit finished();
+        }, Qt::QueuedConnection);
+        jt->disable();
+        jt->go(true);
+    }
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_carbons.cpp]: That was unnecessary check. I removed it.

File: src/xmpp/xmpp-im/xmpp_tasks.cpp [hunk 16]
```diff
@@ -891,19 +892,145 @@ void JT_Message::onGo()
...
+        Subscriber *sbs;
+        int userData;
+    };
     EncryptionHandler *m_encryptionHandler;
+    QHash<QString, QList<SubsData> > subsData;
+    QList<SubsData> subsMData;
+
+    QString genKey(const QString &s1, const QString &s2)
+    {
+        return QString("%1&%2").arg(s1).arg(s2);
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_tasks.cpp]: Fixed. Thanks.

File: src/xmpp/xmpp-im/xmpp_tasks.cpp [hunk 17]
```diff
@@ -891,19 +892,145 @@ void JT_Message::onGo()
...
+    }
+
+    bool processXml(const QDomElement &root, QDomElement &e, Client *c, bool nested)
+    {
+        QString tagName = e.tagName();
+        QString xmlnsStr = e.attribute(QLatin1String("xmlns"));
+        QString key = genKey(tagName, xmlnsStr);
+        if (subsData.contains(key)) {
```
[the remark this answers is not in the document]
Comment (liuch) [reply src/xmpp/xmpp-im/xmpp_tasks.cpp]: You are right. Fixed.

**Comment by Neustradamus** [comment]
It will be merged one day?

**Comment by Neustradamus** [comment]
I think it will be nice to add this in the Google Summer of Code 2023:
- https://wiki.xmpp.org/web/Google_Summer_of_Code_2023

**Comment by Neustradamus** [comment]
@Ri0n: What is missing to have the feature now? After the good work from @mcneb10?

**Comment by Neustradamus** [comment]
A new XEP-0313 version has been released - 1.1.2 (2024-09-18):
- https://xmpp.org/extensions/xep-0313.html
- https://services.w3.org/htmldiff?doc1=https://xmpp.org/extensions/attic/xep-0313-1.1.1.html&doc2=https://xmpp.org/extensions/attic/xep-0313-1.1.2.html

It is good?

Linked to:
- https://github.com/psi-im/iris/pull/73
- https://github.com/psi-im/psi/issues/62

**Comment by Neustradamus** [comment]
Since the good work done by @mcneb10:
- https://github.com/psi-im/iris/pull/73

And after XEP-0313 1.1.2 (2024-09-18), a new XEP-0313 version has been released - 1.1.3 (2025-04-09):
- https://xmpp.org/extensions/xep-0313.html
- https://services.w3.org/htmldiff?doc1=https://xmpp.org/extensions/attic/xep-0313-1.1.2.html&doc2=https://xmpp.org/extensions/attic/xep-0313-1.1.3.html
- https://services.w3.org/htmldiff?doc1=https://xmpp.org/extensions/attic/xep-0313-1.1.1.html&doc2=https://xmpp.org/extensions/attic/xep-0313-1.1.2.html

**Pull request merged**